lib/src/mustachio/parser.dart

```diff
@@ -178,31 +178,50 @@ class MustachioParser {
   /// [_index] should be at the character immediately following the `#`
   /// character which opens a possible section tag.
   _TagParseResult _parseSection({@required invert}) {
-    var result = _parseKey();
-    if (result.type == _KeyParseResultType.notKey) {
+    var parsedKey = _parseKey();
+    if (parsedKey.type == _KeyParseResultType.notKey) {
       return _TagParseResult.notTag;
-    } else if (result == _KeyParseResult.endOfFile) {
+    } else if (parsedKey == _KeyParseResult.endOfFile) {
       return _TagParseResult.endOfFile;
     }
 
-    var children = _parseBlock(sectionKey: result.joinedNames);
+    var children = _parseBlock(sectionKey: parsedKey.joinedNames);
+
+    if (parsedKey.names.length > 1) {
+      // Desugar section with dots into nested sections.
+      //
+      // Given a multi-name section like
+      // `{{#one.two.three}}...{/one.two.three}}`, "one" must be a non-Iterable,
+      // non-bool value. "two" must also be a non-Iterable, non-bool value.
+      // "three" may be any kind of value, resulting in a repeated section,
+      // optional section, or value section. The [children], the parsed AST
+      // inside the section, are the children of the [three] section. The
+      // [three] section is the singular child node of the [two] section, and
+      // the [two] section is the singular child of the [one] section.
+      var section = Section([parsedKey.names.last], children, invert: invert);
+      for (var sectionKey in parsedKey.names.reversed.skip(1)) {
+        section = Section([sectionKey], [section], invert: false);
```

Comment:
Haha I'm glad I took an hour long hike after 2 hours of debugging, just after I realized that multi-name sections were the issue. It was going to be a lot of ugly generated code to get that to work in the rendering side. This solution came to me on the hike. 😁 😁 😁 .

The only shortcoming (I could note this somewhere), is the assumption that each name except the last is a non-Iterable, non-bool value. For example, if an object has an Iterable field like `libraries`, then `{{#libraries.first}}` will just not work. I think this is acceptable, because that same object can just add a getter, `firstLibrary`, etc.

Reply:
Hmm.  I think it's probably fine for the time being, yes, though the example you gave does seem like a reasonable thing someone might try to do.   Probably more of an issue in a generalized implementation.